# Projet Redis 
Fait par @ChloéTellier, @OcéaneGuitton, @FlavieThévenard et @AntoineLucas

### Livrables fourni avec ce rapport :  
 - Un fichier *projet.py*  avec le code brut commenté pour exécuter rapidement notre code.

## Un peu de contexte : c'est quoi Redis ?

Redis, qui sert d'acronyme pour **RE**mote **DI**ctionary **S**erver, est système de gestion de base de données NoSQL (Not Only SQL) de très haute performance, de type clé-valeur, lancé en 2009 par Salvatore Sanfilippo. 


La mémoire de Redis est vertigineuse ce qui permet d’aborder des problématiques BIG DATA avec ce SGBD. Redis est un SGBD scalable, c’est-à-dire qu’il est capable de s’adapter à un changement d’ordre de grandeur de la demande, et plus particulièrement de maintenir ses fonctionnalités et ses performances en cas de forte demande. <br>
Redis est très utilisé par les sites à gros trafic tels que Twitter, The Guardian, Stackoverflow, Netflix, etc. <br>


L'une des grandes différences entre Redis et les autres bases de données NoSQL réside dans les structures de données que Redis fournit. 

Redis accepte tous types de données. En effet, les valeurs peuvent prendre n’importe quelle forme : nombres, chaines, listes, json, xml, html, php, images, vidéos …
Voici quelques types/structures de données manipulables dans Redis :
-        Des chaines de caractères (à une clef est associée une valeur de type chaine de caractère)
-        Des numériques (à une clef est associée une valeur de type numérique)
-        Des listes (plusieurs clefs qui ont des valeurs du même type)
-        Des sets (listes non ordonnées où les clefs ne sont pas répétées, où les clefs sont uniques)
-        Des sets ordonnés (permettant un système de classement)
-        Des hashs (objets contenant des clefs, ces dernières contenant elles-mêmes une liste de valeurs)
-        Des données de géolocalisation
-        Etc.
 


Pour mieux comprendre les différentes structures nous vous invitons à parcourir rapidement le lien [ci-contre](https://redis.io/topics/data-types-intro), et plus généralement la [documentation](https://redis.io/documentation) et le [site](https://redis.io/) de Redis qui ont des documents complets et plutôt clairs.

Une des caractéristiques principales de Redis est de stocker les données en mémoire RAM (Random Access Memory) ou mémoire vive : les données sont stockées sur le court terme. De ce fait, le SGBD Redis a une faible empreinte mémoire.  
Plus particulièrement, Redis permet de gérer les données en « cache » avec le système clé/valeur. 
La mémoire cache est un type de mémoire vive / mémoire RAM à laquelle le microprocesseur peut accéder plus rapidement qu’à la mémoire RAM habituelle. <br>
Gérer les données en cache permet d’accéder à des données momentanément stockées sur la mémoire vive et d’éviter de refaire les mêmes processus de calcul en amont. <br>
Ainsi, Redis permet de gérer des données peu persistantes en présentant la fonctionnalité suivante : il est possible d’attribuer une durée de vie aux données insérées, ce qui permet de mettre en place un système de cache sans se préoccuper de de leur invalidation/disparition. <br>
Du fait des caractéristiques vues précédemment, Redis est fréquemment utilisé pour stocker des données volatiles et fréquemment lues ou encore les sessions utilisateurs. <br>
Par exemple, Redis est utilisé par Twitter ou par des sites marchands pour stocker les informations utilisateurs en mémoire cache.


#### Le type de langage de requête adapté au SGBD Redis est un langage NoSQL orienté clé/valeur.

## Installation de Redis : on met les mains dans le cambouis !

Le site officiel de Redis est assez complet et instinctif. <br>
Redis est en Open Source, il peut donc être utilisé, utilisé et édité librement par tous ses utilisateurs. <br>
Nous allons détailler ici les paramètres d’installation et de prise en main. <br>
Redis peut s’utiliser sur tous les types de systèmes d’exploitation, il est toutefois plus avantageux de l’utiliser sur Linux en comparaison de Windows ou Mac. <br>
Si l’idée est dans un premier temps de vous faire la main sans installer le système sur votre ordinateur nous vous invitons à consulter ce petit [tutoriel](https://try.redis.io/?_ga=2.15710581.2114473602.1604997184-1839395559.1604575298) qui permet de s’essayer aux premières lignes de commandes simples. <br>
<br>
Le site de Redis est configuré de manière assez pédagogique et les informations principales d’installation s’y trouvent, ainsi que toutes les informations de prise en main des commandes (respectivement onglet download et commands).

### Windows

Ce [tutoriel](https://riptutorial.com/fr/redis/example/29962/installation-et-execution-de-redis-server-sous-windows) est très complet et guide toutes les étapes de l’installation de Redis.

### Linux

Il suffit de se rendre à cette [adresse](https://redis.io/topics/quickstart) et de suivre les consignes indiquées dans la partie installation.

Vous serez ensuite prêt en lançant dans un terminal la commande :
```shell 
src/redis-server```

Vous pouvez ensuite lancer des requêtes en lançant le client Redis :
```shell
src/redis-cli```

**Bonus :** pour les utilisateurs Linux, un logiciel interactif visuel pour utiliser Redis existe et est gratuit pour cet OS : [Redis Desktop Manager](https://rdm.dev/pricing).

### Mac OS

Vous pouvez suivre ce [lien](https://medium.com/@petehouston/install-and-config-redis-on-mac-os-x-via-homebrew-eb8df9a4f298) pour installer Redis sur votre machine.

*Nous n'avons pas de machines sous Mac OS et n'avons pas pu vérifier la bonne installation de Redis sur ce système.*


## Installation des librairies utiles dans Python :


Via un terminal de commandes vous devez exécuter les commandes suivantes si vous utilisez [Anaconda](https://www.anaconda.com/).

```shell
conda install redis
conda install pandas
conda install numpy
conda install scipy
conda install seaborn```

Ou bien celles-ci si vous utilisez le gestionnaire de librairies [pip](https://pypi.org/project/pip/).
```shell
pip install redis
pip install pandas
pip install numpy
pip install scipy
pip install seaborn```

Si vous avez des difficultés à installer les modules au sein de votre IDE Python (**Pyzo**, **Spyder**, **Jupyter**, etc.) vous pouvez vous référez à l'aide suivante : [Stackoverflow](https://stackoverflow.com/questions/33897860/python-spyder-conda-install-failure).


## Importation des librairies dans Python :

In [52]:
import redis
import numpy
import scipy
import random
import pprint
import json
import pandas as pd
import seaborn as sns
import time

## Connexion

Comme les autres SGBD, il faut établir une connexion avec Redis pour communiquer avec le SGBD lors de nos requêtes Python.

Ici on fixe les paramètres de connexion (communs pour tous) pour le nom de l'hôte `redis_host` et le numéro de port serveur `redis_port`.

In [53]:
redis_host = "127.0.0.1"
redis_port = 6379

On peut ainsi établir la connexion avec la commande `redis.Redis` en précisant le numéro de la database que l'on veut utiliser. Il y a par défaut 16 databases disponible sur un serveur Redis mais le nombre de bases de données possibles sur Redis est infini : toutefois, il faut s’attendre à des problèmes de performances si on dépasse les 1000 bases de données.

In [4]:
r = redis.Redis(host=redis_host, port=redis_port, db = 0)

## Création de notre base de données 

Nous avons choisi un exemple pour illustrer la force de Redis : la gestion de stock de sushi **(miam)**. <br>
Nous avons fait le choix de créer notre propre jeu de données. <br>
Pour cela nous avons défini un ensemble de clés, et pour chacune de ces clés, nous avons défini un ensemble de « valeurs » / « modalités » possibles stockés dans le format hash de Redis. <br>
Ces hashs peuvent être modifiés n’importe quand de part l’ajout de nouveaux sushis dans la base de données. 

Nous avons choisi de créer notre base de données de sushis en indiquant :
 - La composition d'un sushi selon les ingrédients qui le composent : 73 ingrédients différents, qui peuvent prendre la modalité 'Oui' (le sushi contient cet ingrédient) ou 'Non' (le sushi ne contient pas cet ingrédient).
 - Le stock de ce sushi disponible : généré aléatoirement entre 10 et 10 000.
 - Le nombre de sushis achetés : initialisé à 0.

In [5]:
## Création de la database sushis

random.seed(444)

# Liste des attributs avec 2 modalités : les ingrédients

saumon = ['Oui', 'Non']
saumon_teriyaki = ['Oui', 'Non']
daurade = ['Oui', 'Non']
thon = ['Oui', 'Non']
crevette = ['Oui', 'Non']
poulet = ['Oui', 'Non']
thon_cuit = ['Oui', 'Non']
foie_gras = ['Oui', 'Non']
tofu = ['Oui', 'Non']
truite = ['Oui', 'Non']
hareng = ['Oui', 'Non']
poulpe = ['Oui', 'Non']
boeuf = ['Oui', 'Non']
chair_de_crabe = ['Oui', 'Non']
oeufs_de_saumon = ['Oui', 'Non']
avocat = ['Oui', 'Non']
fromage = ['Oui', 'Non']
oeuf = ['Oui', 'Non']
gingembre = ['Oui', 'Non']
wasabi = ['Oui', 'Non']
sauce_salee = ['Oui', 'Non']
sauce_sucree = ['Oui', 'Non']
sauce_sucreesalee = ['Oui', 'Non']
mayonnaise_classique = ['Oui', 'Non']
mayonnaise_teriyaki = ['Oui', 'Non']
mayonnaise_japonaise = ['Oui', 'Non']
mayonnaise_spicy = ['Oui', 'Non']
mayonnaise_ponzu = ['Oui', 'Non']
sauce_teriyaki = ['Oui', 'Non']
sauce_satay_aux_cacahuetes = ['Oui', 'Non']
sauce_epicee = ['Oui', 'Non']
mangue = ['Oui', 'Non']
carotte = ['Oui', 'Non']
anis = ['Oui', 'Non']
poivre_rose = ['Oui', 'Non']
cannelle = ['Oui', 'Non']
cardamome = ['Oui', 'Non']
curcuma = ['Oui', 'Non']
feve = ['Oui', 'Non']
edamame = ['Oui', 'Non']
macis = ['Oui', 'Non']
maniguette = ['Oui', 'Non']
paprika = ['Oui', 'Non']
piment = ['Oui', 'Non']
poivre = ['Oui', 'Non']
safran = ['Oui', 'Non']
sumac = ['Oui', 'Non']
persil = ['Oui', 'Non']
herbe_de_provence = ['Oui', 'Non']
sesame = ['Oui', 'Non']
feuilles_de_riz = ['Oui', 'Non']
menthe = ['Oui', 'Non']
coriandre = ['Oui', 'Non']
chou = ['Oui', 'Non']
ciboulette = ['Oui', 'Non']
pomme = ['Oui', 'Non']
celeri_rave = ['Oui', 'Non']
aneth = ['Oui', 'Non']
baies_roses = ['Oui', 'Non']
prune = ['Oui', 'Non']
betterave = ['Oui', 'Non']
noix_de_coco = ['Oui', 'Non']
citron_vert = ['Oui', 'Non']
citron_jaune = ['Oui', 'Non']
dattes = ['Oui', 'Non']
laitue = ['Oui', 'Non']
roquette = ['Oui', 'Non']
concombre = ['Oui', 'Non']
poivrons = ['Oui', 'Non']
asperge = ['Oui', 'Non']
oignons_crus = ['Oui', 'Non']
oignons_caramelises = ['Oui', 'Non']
oignons_frits = ['Oui', 'Non']

# 2**73 = 9.5*(10**21) combinaisons possibles de sushis

sushis = []

tps1 = time.perf_counter()

for i in range(0, 100000):
    saumon_choice = random.choice(saumon)
    saumon_teriyaki_choice = random.choice(saumon_teriyaki)
    daurade_choice = random.choice(daurade)
    thon_choice = random.choice(thon)
    crevette_choice = random.choice(crevette)
    poulet_choice = random.choice(poulet)
    thon_cuit_choice = random.choice(thon_cuit)
    foie_gras_choice = random.choice(foie_gras)
    tofu_choice = random.choice(tofu)
    truite_choice = random.choice(truite)
    hareng_choice = random.choice(hareng)
    poulpe_choice = random.choice(poulpe)
    boeuf_choice = random.choice(boeuf)
    chair_de_crabe_choice = random.choice(chair_de_crabe)
    oeufs_de_saumon_choice = random.choice(oeufs_de_saumon)
    avocat_choice = random.choice(avocat)
    fromage_choice = random.choice(fromage)
    oeuf_choice = random.choice(oeuf)
    gingembre_choice = random.choice(gingembre)
    wasabi_choice = random.choice(wasabi)
    sauce_salee_choice = random.choice(sauce_salee)
    sauce_sucree_choice = random.choice(sauce_sucree)
    sauce_sucreesalee_choice = random.choice(sauce_sucreesalee)
    mayonnaise_classique_choice = random.choice(mayonnaise_classique)
    mayonnaise_teriyaki_choice = random.choice(mayonnaise_teriyaki)
    mayonnaise_japonaise_choice = random.choice(mayonnaise_japonaise)
    mayonnaise_spicy_choice = random.choice(mayonnaise_spicy)
    mayonnaise_ponzu_choice = random.choice(mayonnaise_ponzu)
    sauce_teriyaki_choice = random.choice(sauce_teriyaki)
    sauce_satay_aux_cacahuetes_choice = random.choice(sauce_satay_aux_cacahuetes)
    sauce_epicee_choice = random.choice(sauce_epicee)
    mangue_choice = random.choice(mangue)
    carotte_choice = random.choice(carotte)
    anis_choice = random.choice(anis)
    poivre_rose_choice = random.choice(poivre_rose)
    cannelle_choice = random.choice(cannelle)
    cardamome_choice = random.choice(cardamome)
    curcuma_choice = random.choice(curcuma)
    feve_choice = random.choice(feve)
    edamame_choice = random.choice(edamame)
    macis_choice = random.choice(macis)
    maniguette_choice = random.choice(maniguette)
    paprika_choice = random.choice(paprika)
    piment_choice = random.choice(piment)
    poivre_choice = random.choice(poivre)
    safran_choice = random.choice(safran)
    sumac_choice = random.choice(sumac)
    persil_choice = random.choice(persil)
    herbe_de_provence_choice = random.choice(herbe_de_provence)
    sesame_choice = random.choice(sesame)
    feuilles_de_riz_choice = random.choice(feuilles_de_riz)
    menthe_choice = random.choice(menthe)
    coriandre_choice = random.choice(coriandre)
    chou_choice = random.choice(chou)
    ciboulette_choice = random.choice(ciboulette)
    pomme_choice = random.choice(pomme)
    celeri_rave_choice = random.choice(celeri_rave)
    aneth_choice = random.choice(aneth)
    baies_roses_choice = random.choice(baies_roses)
    prune_choice = random.choice(prune)
    betterave_choice = random.choice(betterave)
    noix_de_coco_choice = random.choice(noix_de_coco)
    citron_vert_choice = random.choice(citron_vert)
    citron_jaune_choice = random.choice(citron_jaune)
    dattes_choice = random.choice(dattes)
    laitue_choice = random.choice(laitue)
    roquette_choice = random.choice(roquette)
    concombre_choice = random.choice(concombre)
    poivrons_choice = random.choice(poivrons)
    asperge_choice = random.choice(asperge)
    oignons_crus_choice = random.choice(oignons_crus)
    oignons_caramelises_choice = random.choice(oignons_caramelises)
    oignons_frits_choice = random.choice(oignons_frits)
    stock = random.randint(10,10000)
    nb_achat = 0
    sushi = {'id':i, 'saumon':saumon_choice, 'saumon_teriyaki':saumon_teriyaki_choice, 'daurade':daurade_choice, 
             'thon':thon_choice, 'crevette':crevette_choice, 'poulet':poulet_choice, 'thon_cuit':thon_cuit_choice, 
             'foie_gras':foie_gras_choice,'tofu':tofu_choice, 'truite':truite_choice, 'hareng':hareng_choice, 
             'poulpe':poulpe_choice, 'boeuf':boeuf_choice, 'chair_de_crabe':chair_de_crabe_choice, 
             'oeufs_de_saumon':oeufs_de_saumon_choice, 'avocat':avocat_choice, 'fromage':fromage_choice, 
             'oeuf':oeuf_choice, 'gingembre':gingembre_choice,'wasabi':wasabi_choice, 'sauce_salee':sauce_salee_choice, 
             'sauce_sucree':sauce_sucree_choice,'sauce_sucreesalee':sauce_sucreesalee_choice, 
             'mayonnaise_classique':mayonnaise_classique_choice, 'mayonnaise_teriyaki':mayonnaise_teriyaki_choice, 
             'mayonnaise_japonaise':mayonnaise_japonaise_choice, 'mayonnaise_spicy':mayonnaise_japonaise_choice, 
             'mayonnaise_ponzu':mayonnaise_ponzu_choice, 'sauce_teriyaki':sauce_teriyaki_choice, 
             'sauce_satay_aux_cacahuetes':sauce_satay_aux_cacahuetes_choice, 'sauce_epicee':sauce_epicee_choice, 
             'mangue': mangue_choice, 'carotte':carotte_choice, 'anis':anis_choice,'poivre_rose':poivre_rose_choice, 
             'cannelle':cannelle_choice, 'cardamome':cardamome_choice, 'curcuma':curcuma_choice, 
             'feve':feve_choice, 'edamame':edamame_choice, 'macis':macis_choice, 'maniguette':maniguette_choice, 
             'paprika':paprika_choice, 'piment':piment_choice, 'poivre':poivre_choice, 'safran':safran_choice, 
             'sumac': sumac_choice, 'persil':persil_choice, 'herbe_de_provence':herbe_de_provence_choice, 
             'sesame':sesame_choice, 'feuilles_de_riz':feuilles_de_riz_choice, 'menthe':menthe_choice, 
             'coriandre':coriandre_choice, 'chou':chou_choice, 'ciboulette':ciboulette_choice, 'pomme':pomme_choice, 
             'celeri_rave':celeri_rave_choice, 'aneth':aneth_choice, 'baies_roses':baies_roses_choice, 
             'prune':prune_choice, 'betterave':betterave_choice, 'noix_de_coco':noix_de_coco_choice, 
             'citron_vert':citron_vert_choice, 'citron_jaune':citron_jaune_choice, 'dattes':dattes_choice, 
             'laitue':laitue_choice, 'roquette':roquette_choice, 'concombre':concombre_choice, 
             'poivrons':poivrons_choice, 'asperge':asperge_choice, 'oignons_crus':oignons_crus_choice, 
             'oignons_caramelises':oignons_caramelises_choice, 'oignons_frits':oignons_frits_choice,'stock':stock, 'nb_achat':nb_achat
            }
    
    sushis.append(sushi)

tps2 = time.perf_counter()
print(tps2 - tps1)

4.5599060400000155


Le temps pour créer une liste de 100 000 sushis différents est d'environ 4.56 secondes.

## On ajoute ensuite nos données générées dans Redis 

Une des forces de redis est l'utilisation de pipeline pour réaliser les requêtes, nous y reviendrons en fin de rapport. Pour plus d'informations, rendez-vous [ici](https://redis.io/topics/pipelining).

In [6]:

with r.pipeline() as pipe: 
    for i in range(len(sushis)):
        name = str("sushi:")+str(i)
        r.hmset(name,sushis[i])
        

<ipython-input-6-4b3353b65ed5>:4: DeprecationWarning: Redis.hmset() is deprecated. Use Redis.hset() instead.
  r.hmset(name,sushis[i])


Le durée pour ajouter une liste de 100 000 sushis différents à Redis est d'environ 25 secondes.

## Nous sommes prêts pour faire quelques requêtes et explorer les possibilités de Redis !

On peut ensuite effectuer quelques requêtes simples avec les objets **hashs** que nous utiliserons au cours de notre projet.
Pour plus d'informations sur les requêtes sur les types d'objets **hashs**, rendez-vous [ici](https://redis.io/commands#hash).

Ainsi, comme exemple, nous avons nos clés avec la dénomination *sushi:i* avec i le numéro du sushi allant de 1 à 100 000. Nous pouvons alors récupérer au sein de chaque clé différentes informations stockées sous une forme similaire à un dictionnaire.

In [48]:
stock_exemple_propre = int(r.hget('sushi:3564','stock').decode())
stock_exemple_bytes = r.hget('sushi:3564','stock')
print("En écriture propre : ",stock_exemple_propre," / ","En écriture Bytes : ",stock_exemple_bytes)

En écriture propre :  1894  /  En écriture Bytes :  b'1894'


Ainsi, on récupère le stock (désigné par `stock` du *sushi:3*.
Notez l'utilisation de `int()` et `.decode()` pour récupérez sous un format plus "agréable" les requêtes de Redis qui sont retournées par défaut au format **bytes**.

## Création de fonctions utiles

Nous allons ensuite définir quelques fonctions utiles pour réaliser quelques requêtes.
Pour cela nous aurons besoin de définir certains messages d'erreurs.

In [50]:
# On définit nos messages d'erreur, que l'on utilisera pour prévenir l'utilisateur selon différents critères 

# Plus de stock pour le sushi désiré
class OutOfStockError(Exception):
    """Notre magasin de sushis n'a plus du tout de stock pour le sushi désiré"""
    
# Stock rempli pour le sushi désiré : stock fixé à 10 000 maximum
class TooMuchStockError(Exception):
    """Notre magasin de sushis n'a plus de place pour stocker le sushi d'intérêt, nous n'avons rien pu ajouter
    au stock"""

# Le client désirait acheter trop de sushis et a fini la fin du stock
class TooMuchDemandError(Exception):
    """Le client souhaitait acheter plus de sushis qu'il n'y en avait de disponibles, il a fini le stock
    et n'a pas eu autant de sushis qu'il l'esperait"""

# Nous avons restocké au maximum les 10 000 sushis
class NoPlaceAvailableError(Exception):
    """Notre magasin n'a pas assez de place pour stocker tous les sushis que l'on souhaitait,
    seuls certains ont pu être ajoutés au stock"""


En outre, Redis propose des fonctionnalités d’incrémentation et de décrémentation intéressantes : il est possible d’incrémenter et de décrémenter facilement et simultanément un grand nombre de valeurs. Ce qui fait que ce SGBD est bien adapté pour la gestion de stock et la vente en temps réel.

Nous allons maintenant créer la fonction `buyitem`. Celle-ci permet d'acheter un certain nombre de sushis (argument `count`) d'un certain type (argument `sushi_id`) : cela diminue-donc le stock de sushis, mais augmente le nombre de sushis achetés.

In [42]:
def buyitem(r, sushi_id,count):

    with r.pipeline() as pipe:
        
        while True:

            pipe.watch(sushi_id)
            nleft = int(r.hget(sushi_id,'stock').decode())
            # On utilise le int et decode pour convertir l'output bytes de Redis en nombre int
            
            # On réalise l'opération d'achat normalement s'il y a assez de sushis en stock
            if nleft > count :
                pipe.multi() # On débute l'enregistrement des étapes qu'on veux réaliser sur le serveur redis
                pipe.hincrby(sushi_id, 'stock', -count) # Le stock diminue ...
                pipe.hincrby(sushi_id, 'nb_achat', count) # ... Et le nombre de sushis achetés augmente
                pipe.execute() # On transfert l'ensemble des étapes réalisée depuis multi sur notre serveur Redis
                break
            
            # Si le nombre de sushis n'est pas suffisant pour satisfaire la demande, on achète tout ce qu'il reste
            elif nleft < count and nleft > 0:
                pipe.multi()
                pipe.hset(sushi_id,'stock',0)
                pipe.hincrby(sushi_id,'nb_achat',nleft)
                pipe.execute()
                
                raise TooMuchDemandError(f'Désolé, vous avez acheté tout les {sushi_id} restants mais il n''y en avait pas autant que vous le vouliez')

            # S'il n'y a plus du tout de ce type de sushis ...
            else:
                pipe.unwatch()
                raise OutOfStockError(f"Sorry, {sushi_id} is out of stock!")

    return None

Testons maintenant cette fonction : et si on achetait 60 sushis de type 5 ?

In [43]:
buyitem(r,"sushi:5",60)
r.hmget("sushi:5","stock","nb_achat")

[b'6045', b'180']

Re-exécutez la cellule précédente : on a modifié les valeurs de stock et de la quantité d'achat sur notre database.

Nous allons maintenant créer la fonction `restock`. Celle-ci permet au magasin de se ré-approvisionner d'un certain nombre (argument `stock_count`) de sushis d'un certain type (argument `sushi_id`) : le stock de ce type de sushis augmente ainsi.
Arbitrairement, nous avons choisi qu'un magasin ne pouvait stocker que 10 000 sushis d'un même type (c'est déjà pas mal).

In [44]:
def restock(r,sushi_id,stock_count):
    
    with r.pipeline() as pipe:
        
        while True:
        
            pipe.watch(sushi_id)
            nleft_restock = int(r.hget(sushi_id,'stock').decode())

            # S'il y a suffisamment de places pour stocker les sushis, on achète le nombre demandé
            if nleft_restock + stock_count < 10000:
                pipe.multi()
                pipe.hincrby(sushi_id, 'stock', stock_count) # On modifie le stock sur notre database Redis
                pipe.execute()
                break
            
            # S'il n'y a pas suffisamment de place pour stocker tous les nouveaux sushis, on achète seulement ce que l'on peut
            elif nleft_restock < 10000 and nleft_restock + stock_count > 10000:
                pipe.multi()
                pipe.hset(sushi_id,'stock',10000)
                pipe.execute()                
                raise NoPlaceAvailableError(f"Désolé, nous avons déjà remis le stock de {sushi_id} au maximum")

            # Si le stock est déjà plein, cela ne sert à rien d'acheter davantage de sushis !
            # (et de toute façon on ne peut pas les stocker)
            else:
                pipe.unwatch()
                raise TooMuchStockError(f"Sorry, {sushi_id} is already full of stock!")
            
    return None

Testons maintenant cette fonction : et si on augmentait le stock du sushi 9 de 400 ?

In [30]:
restock(r,'sushi:9', 400)
r.hmget("sushi:9",'stock','nb_achat')

[b'9742', b'0']

En répétant la cellule précédente plusieurs fois, on augmente la quantité stockée du sushi 9 comme précédemment.

Nous allons maintenant créer la fonction `item_info`. Celle-ci permet de récupérer toutes les informations relatives à nos ventes de sushis : stock, vente. Ces informations seront ensuite stockées dans un dataframe [Pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html).

In [45]:
def item_info():
    
    info_get = [] # Tableau qui contiendra les données récupérées
    
    for i in range(len(sushis)):
        name = str("sushi:")+str(i)
        stock = int(r.hget(name,'stock').decode())
        nb_achat = int(r.hget(name,'nb_achat').decode())
        info_add = [stock, nb_achat]               
        info_get.append(info_add)
        
    # On convertit notre liste de liste en dataframe
    item_df = pd.DataFrame(info_get,
                        columns=['stock','nb_achat'],
                        index = [str("sushi:")+str(i) for i in range(len(sushis))])
    
    return(item_df)

In [32]:
liste_sushi = item_info()
liste_sushi

stock  nb_achat
sushi:0       4846         0
sushi:1       1596         0
sushi:2       6646         0
sushi:3       7786         0
sushi:4       9207         0
...            ...       ...
sushi:99995   9550         0
sushi:99996   5211         0
sushi:99997   8203         0
sushi:99998   1215         0
sushi:99999   5521         0

[100000 rows x 2 columns]

La fonction `item_ingredients`, elle, permet de récupérer les ingrédients de nos différents types de sushis (73 variables donc, correspondant à l'absence ou à la présence de chacun des ingrédients). Ces informations seront ensuite stockées dans un dataframe Pandas.

In [33]:
def item_ingredient():
    
    ingredient_get = []
    
    for i in range(len(sushis)):
        
        name = str("sushi:")+str(i)
        ingredient_add = []
        
        name_ingredient = ['saumon', 'saumon_teriyaki', 'daurade', 'thon', 'crevette', 
                                            'poulet', 'thon_cuit', 'foie_gras', 'tofu', 'truite', 'hareng', 
                                            'poulpe', 'boeuf', 'chair_de_crabe', 'oeufs_de_saumon', 'avocat', 
                                            'fromage', 'oeuf', 'gingembre', 'wasabi', 'sauce_salee', 
                                            'sauce_sucree', 'sauce_sucreesalee', 'mayonnaise_classique', 
                                            'mayonnaise_teriyaki', 'mayonnaise_japonaise', 'mayonnaise_spicy', 
                                            'mayonnaise_ponzu', 'sauce_teriyaki', 'sauce_satay_aux_cacahuetes',
                                            'sauce_epicee', 'mangue', 'carotte', 'anis', 'poivre_rose', 
                                            'cannelle', 'cardamome', 'curcuma', 'feve', 'edamame', 'macis', 
                                            'maniguette', 'paprika', 'piment', 'poivre', 'safran', 'sumac', 
                                            'persil', 'herbe_de_provence', 'sesame', 'feuilles_de_riz', 
                                            'menthe', 'coriandre', 'chou', 'ciboulette', 'pomme', 
                                            'celeri_rave', 'aneth', 'baies_roses', 'prune', 'betterave', 
                                            'noix_de_coco', 'citron_vert', 'citron_jaune', 'dattes', 
                                            'laitue', 'roquette', 'concombre', 'poivrons', 'asperge', 
                                            'oignons_crus', 'oignons_caramelises', 'oignons_frits']

        # La boucle permet de rentrer l'ensemble des ingrédients dans le tableau
        for i in range(len(name_ingredient)):
            ingredient = r.hget(name, name_ingredient[i]).decode() # On récupère les infos depuis la database Redis
            ingredient_add.append(ingredient)
        
        ingredient_get.append(ingredient_add)
        
    ingredient_df = pd.DataFrame(ingredient_get,
                        columns=name_ingredient,
                        index = [str("sushi:")+str(i) for i in range(len(sushis))])
    
    return(ingredient_df)

In [34]:
ingredient_df = item_ingredient()
ingredient_df

saumon saumon_teriyaki daurade thon crevette poulet thon_cuit  \
sushi:0        Non             Non     Oui  Non      Non    Non       Non   
sushi:1        Oui             Oui     Non  Oui      Non    Oui       Non   
sushi:2        Non             Oui     Oui  Oui      Non    Oui       Non   
sushi:3        Oui             Non     Non  Non      Non    Oui       Non   
sushi:4        Oui             Non     Non  Oui      Non    Non       Non   
...            ...             ...     ...  ...      ...    ...       ...   
sushi:99995    Non             Oui     Oui  Non      Non    Oui       Oui   
sushi:99996    Non             Oui     Non  Non      Non    Oui       Oui   
sushi:99997    Oui             Non     Non  Non      Oui    Non       Oui   
sushi:99998    Oui             Non     Non  Non      Non    Oui       Non   
sushi:99999    Non             Non     Non  Oui      Oui    Oui       Oui   

            foie_gras tofu truite  ... citron_jaune dattes laitue roquette  \
sushi:0           Non  Non    Non  ...          Oui    Non    Non      Oui   
sushi:1           Oui  Non    Non  ...          Non    Non    Non      Non   
sushi:2           Non  Non    Non  ...          Oui    Non    Oui      Non   
sushi:3           Non  Non    Non  ...          Oui    Oui    Oui      Non   
sushi:4           Non  Non    Non  ...          Oui    Oui    Non      Oui   
...               ...  ...    ...  ...          ...    ...    ...      ...   
sushi:99995       Non  Non    Oui  ...          Non    Non    Non      Oui   
sushi:99996       Oui  Oui    Non  ...          Non    Oui    Non      Non   
sushi:99997       Oui  Oui    Non  ...          Non    Oui    Non      Oui   
sushi:99998       Oui  Oui    Oui  ...          Non    Oui    Non      Non   
sushi:99999       Oui  Non    Non  ...          Oui    Non    Non      Oui   

            concombre poivrons asperge oignons_crus oignons_caramelises  \
sushi:0           Non      Non     Oui          Oui                 Oui   
sushi:1           Non      Oui     Oui          Non                 Oui   
sushi:2           Non      Non     Oui          Non                 Non   
sushi:3           Oui      Non     Oui          Oui                 Oui   
sushi:4           Oui      Oui     Non          Oui                 Non   
...               ...      ...     ...          ...                 ...   
sushi:99995       Non      Oui     Non          Oui                 Non   
sushi:99996       Non      Oui     Oui          Oui                 Oui   
sushi:99997       Non      Oui     Oui          Oui                 Oui   
sushi:99998       Non      Non     Non          Oui                 Oui   
sushi:99999       Oui      Oui     Non          Oui                 Non   

            oignons_frits  
sushi:0               Oui  
sushi:1               Oui  
sushi:2               Non  
sushi:3               Non  
sushi:4               Non  
...                   ...  
sushi:99995           Non  
sushi:99996           Oui  
sushi:99997           Oui  
sushi:99998           Non  
sushi:99999           Non  

[100000 rows x 73 columns]

La fonction `sushi_interet`, définie ci-dessous, est très pratique pour les clients. Ceux-ci peuvent entrer en arguments une liste des ingrédients qu'ils aiment dans les sushis, et la fonction leur retourne tous les sushis de la carte qui contiennent l'ensemble de ces ingrédients récupérés précédemment sur la database redis.

In [51]:
def sushi_interet(r,liste_ingredients,ingredient_df):
        
    sushi_interessant = [] # Liste qui stockera les numéros de tous les sushis contenant tous les ingrédients de liste_ingredients
    for j in range(len(ingredient_df)): # On prend tous les sushis successivement ...
        if all(ingredient_df.iloc[j][liste_ingredients]=="Oui"): # ... et on regarde s'ils contiennent tous les ingrédients demandés
            sushi_interessant.append(j)
            
    phrase = "Vous pourriez apprécier les sushis suivants : "
    for i in range(len(sushi_interessant)):
        phrase += "sushi:"+str(sushi_interessant[i])+" "

    print(phrase)
    return(sushi_interessant)

Faisons un essai : je suis un client qui aime le saumon, l'avocat, le thon, la roquette, la cannelle, la cardamone, le curcuma, les fèves, les edamame, le paprika, le piment, le poivre et le safran, et je souhaiterais un sushi qui mélange toutes ces saveurs (rien que ça).

In [47]:
liste_ingredients = ['saumon', 'avocat', 'thon', 'roquette', 'cannelle', 'cardamome', 'curcuma',
                     'feve', 'edamame', 'paprika', 'piment', 'poivre', 'safran']
sushi_interet(r,liste_ingredients,ingredient_df)

Vous pourriez apprécier les sushis suivants : sushi:1083 sushi:1872 sushi:11616 sushi:12887 sushi:16489 sushi:20978 sushi:22943 sushi:24880 sushi:25332 sushi:37258 sushi:39997 sushi:59722 sushi:73204 sushi:75115 sushi:80214 sushi:84255 sushi:93868 sushi:94754 


[1083,
 1872,
 11616,
 12887,
 16489,
 20978,
 22943,
 24880,
 25332,
 37258,
 39997,
 59722,
 73204,
 75115,
 80214,
 84255,
 93868,
 94754]

## Bilan de l'outil SGBD Redis

Redis présente de nombreux **avantages** :
- Une prise en main très facile et une super documentation : tout est renseigné, on comprend assez vite le fonctionnement de ce SGBD.
- Il est Open Source et accessible facilement (ce n'est toutefois pas le cas de l'environnement autour, car le projet a été repris par une entreprise : [redislabs](https://redislabs.com/)).
- Il est très flexible (Redis permet de manipuler facilement à peu près n’importe quel type de données).
- Il permet également une très bonne portabilité (les données, et plus particulièrement les clefs/valeurs stockées peuvent être facilement déplacées d’un système à l’autre sans avoir à réécrire du code).

**Toutefois :** <br>
- Il est difficile de se projeter dans l'usage de ce SGBD : il est très différent des autres et assez spécifique. Même si on peux l'utiliser pour tout usage, il sera surtout performant dans des tâches spécifiques.
- La persistance des données n’étant pas optimale sur Redis, ce SGBD est souvent utilisé comme base de données secondaire, en complément d’un autre type de base de données.
- Il est recommandé de ne pas stocker sur Redis des informations sensibles ou des informations que l’on n’a pas conservé ailleurs. Toutefois, afin d’assurer la persistance des données (et d’éviter la perte de données en cas d’incident tel qu’un crash), Redis offre la possibilité de capturer l’état de la base de données à un temps donné dans un fichier .rdb.
- L’accès aux données n’est pas sécurisé. Une personne connectée sur le serveur pourra exploiter toutes les données Redis, il n’est donc pas conseillé de stocker des données sensibles sur ce SGBD.